### PR TITLE
IHS-18609 IHS-18612 BTF blacklisting rules

### PR DESCRIPTION
**Pull request by jefffurlong**
IHS-18609 and IHS-18612 stories.

### CODE REVIEW

**Comment by R-Hallett-introhive** [comment]
Tested the new code changes locally and saw all the diag entries in my local db. 
The Whitelisted files were created and all the blocked contacts were removed from them.
I also reviewed the code changes for these PR's. The code changes looked really good and don't have much to suggest. My only suggestion would be to add more description on what changed and add in some testing steps for the PR.

File: voip-btf/Blacklist.cs [hunk 1]
```diff
@@ -168,76 +171,154 @@ public static string GetJsonStream(string url, string username, string password)
...
         }
-        
+
+        public static void FetchBlacklists()
+        {
+            //Print Process Start Retrieving Blacklists
+            Log.Information("Blacklist Start: ");
+
+            //Retrieve Blacklisted Domains
+            string pingpong = Blacklist.PostJsonStream(Blacklist.settings.url + "diag/ping.json",
```
Comment (SteveoMorris) [inline voip-btf/Blacklist.cs]: Could you re-order the calls to the diag/ping endpoint like the following:
1) Call diag/ping for "Start retrieving blacklist domain"
2) fetch blacklist domains
3) Call diag/ping for "Finished retrieving blacklist domain..."
4) Call diag/ping for "Start retrieving blacklist contacts"
5) fetch blacklist contacts
6) Call diag/ping for "Finished retrieving blacklist contacts..."

The way it works now, if it crashes we will get the start msgs and not the end msgs.  Doing it in the order above will help troubleshoot if something goes wrong

File: voip-btf/Blacklist.cs [hunk 2]
```diff
@@ -168,76 +171,154 @@ public static string GetJsonStream(string url, string username, string password)
...
+                    "Found " + Blacklist.TotalContactEntries + " Blacklist contact entries", Blacklist.uniqueid);
+
+            }
+
+
+            //Print Process Finished Retrieving Blacklists
+            Log.Information("Blacklist End: ");
+
+            //Print Process End Time
+            pingpong = Blacklist.PostJsonStream(Blacklist.settings.url + "diag/ping.json",
```
Comment (SteveoMorris) [inline voip-btf/Blacklist.cs]: Do we still need this?  Is this left over from the original story?  We should be making a call to diag/ping at the end of the entire application run saying "Exiting Application"

File: voip-btf/Blacklist.cs [hunk 3]
```diff
@@ -300,31 +382,284 @@ public static async Task QueryContactsWithPhoneNumberNotEmpty()
...
+                {
+                    for (int j = 0; j < Convert.ToInt32(TotalDomainEntries); j++)
+                    {
+
+                        bDeleteRecord = savedList[i].Email.Contains(domains[j]);
+
+                        if (bDeleteRecord)
+                        {
+                            Log.Information("Deleting Record: " + savedList[i].Email);
+                            savedList.RemoveAt(i);
```
Comment (SteveoMorris) [inline voip-btf/Blacklist.cs]: Does this work?  From the docs on RemoveAt() it says that when this method is called, the collection is re-ordered.  So as we are looping through the savedList collection using i as the counter, when we do a savedList.Remove(3), item 4 now becomes item 3.

From the docs:
"When you call RemoveAt to remove an item, the remaining items in the list are renumbered to replace the removed item. For example, if you remove the item at index 3, the item at index 4 is moved to the 3 position. In addition, the number of items in the list (as represented by the Count property) is reduced by 1."

File: voip-btf/Blacklist.cs [hunk 4]
```diff
@@ -300,31 +382,284 @@ public static async Task QueryContactsWithPhoneNumberNotEmpty()
...
             {
-                contactsBuffer = buffer.ToArray();
+                if (bDomainExist)
+                {
+                    for (int j = 0; j < Convert.ToInt32(TotalDomainEntries); j++)
+                    {
+
+                        bDeleteRecord = savedList[i].Email.Contains(domains[j]);
+
+                        if (bDeleteRecord)
```
Comment (SteveoMorris) [inline voip-btf/Blacklist.cs]: Do we need this boolean variable?  Can we just change the if statement to:
if(savedList[i].Email.Contains(domains[j]))

It doesn't look like we do anything with the variable (other than reset it)

File: voip-btf/Blacklist.cs [hunk 5]
```diff
@@ -300,31 +382,284 @@ public static async Task QueryContactsWithPhoneNumberNotEmpty()
...
+
+                    }
+                }
+                
+                if (bContactExist)
+                {
+                    for (int k = 0; k < Convert.ToInt32(TotalContactEntries); k++)
+                    {
+
+                        bDeleteRecord = savedList[i].Email.Contains(contacts[k]);
```
Comment (SteveoMorris) [inline voip-btf/Blacklist.cs]: Same comment as above regarding this variable / if statement

File: voip-btf/Blacklist.cs [hunk 6]
```diff
@@ -300,31 +382,284 @@ public static async Task QueryContactsWithPhoneNumberNotEmpty()
...
+
+                        if (bDeleteRecord)
+                        {
+                            Log.Information("Deleting Record: " + savedList[i].Email);
+                            savedList.RemoveAt(i);
```
Comment (SteveoMorris) [inline voip-btf/Blacklist.cs]: Same as above

File: voip-btf/Blacklist.cs [hunk 7]
```diff
@@ -270,6 +351,7 @@ public static bool InitSalesforce()
 
         }
         
+        //Query method to retrieve phone numbers from Salesforce CRM
```
Comment (SteveoMorris) [inline voip-btf/Blacklist.cs]: The query in this method searches where Phone != null OR MobilePhone != null.  Do we need to add the AssitantPhone, OtherPhone and HomePhone != null to that where clause?

File: voip-btf/Blacklist.cs [hunk 8]
```diff
@@ -56,6 +57,11 @@ public static class settings
         public static string uniqueid;
 
         public static string[] files;
+        public static List<String> domains;
+        public static List<String> contacts;
+        public static int noRecordsDeleted;
```
Comment (SteveoMorris) [inline voip-btf/Blacklist.cs]: Can you rename this to nbrRecordsDeleted or NumberRecordsDeleted.  When I was reading the code, I though it was a boolean flag saying "no records deleted".

File: voip-btf/Blacklist.cs (hunk 3 above)
Comment (SteveoMorris) [reply voip-btf/Blacklist.cs]: actually this is going through the file in reverse order, so it isn't an issue

**Pull request merged**